doc/devs.rst

```diff
@@ -0,0 +1,321 @@
+  - Each peer can publish to each topic. No central coordinator or root node
+    exists.
+  - Broker minimizes messages in the network with (application-layer) multicast
+    in order to scale effectively.
+
+Non-Goals
+  - Dynamic connection management and auto-balancing of a distribution tree.
+    While Broker generally follows a P2P philosophy, we still wish to give users
+    full control over Broker's topology. Also, Broker does not target
+    Internet-wide deployments with very high churn rates and unreliable nodes.
+
+Peers
+~~~~~
+
+Each Broker peer in the network has:
+
+- A globally unique ID. Currently, we use ``caf::node_id`` to identify peers.
+  CAF computes this ID automatically by combining a 160-bit hash value (based on
+  a seed plus various node-specific information) with the OS-specific process
+  ID.
+- Local subscriptions. Endpoints combine the subscriptions of all local
+  subscribers to a single subscription list. The endpoint removes all redundant
+  entries from this list. For example, an endpoint with local subscribers for
+  ``/zeek/event/foo``, ``/zeek/event/bar``, and ``/zeek/event`` is only going to
+  store ``/zeek/event``. This subscription automatically includes the other two
+  due to the prefix matching.
+- A logical clock (Lamport timestamps). This 64-bit integer enables detection of
+  repeated messages and ordering of events. Whenever a peer sends a message to
+  others, it increments its logical time and includes the current value in the
+  message. This timestamp is crucial for detecting outdated or repeated
+  subscriptions in the `Subscription Flooding`_.
+- A routing table with paths to *all* known peers in the network as well as the
+  subscriptions of each peer.
+
+Subscription Flooding
+~~~~~~~~~~~~~~~~~~~~~
+
+Whenever the local subscriptions change, a peer sends a *subscription* message
+to all of its direct connections. When establishing a new peering relation, the
+handshake also includes the *subscription* message.
+
+The subscription message consists of:
+
+#. A ``peer_id_list`` for storing the path of this message. Initially, this list
+   only contains the ID of the sender.
+#. A ``filter_type`` (list of topics) that denotes the current subscriptions.
+#. A 64-bit (unsigned) timestamp. This is the logical time of the sender for
+   this event.
+
+Whenever receiving a *subscription* message (this ultimately calls
+``handle_subscription`` in ``include/broker/alm/peer.hh``), a peer first checks
+whether the path already contains its ID, in which case it discards the message
+since it contains a loop.
+
+If a peer sees the sender (the first entry in the path) for the first time, it
+stores the subscription (and the path). Otherwise, it checks the timestamp of
+the message:
+
+- If the timestamp is *less* than the last timestamp, a peer simply drops the
+  outdated message.
+- If the timestamp is *equal* to the last timestamp, a peer stores the new path.
```

Comment:
Does same timestamp mean it's a message we already saw before? if so, wouldn't we already be in the path, and hence have dropped the message with the first check above? And what's the updated path giving us?